### PR TITLE
Feature/multigroup part10

### PR DESCRIPTION
**Pull request by whaeck**
This adds the last few remaining records for the multigroup neutron and photon files:
- nubar
- nu sigma_f
- chi vector
- chi matrix

The only few remaining things to finalise are the constructor of the MultigroupTable to use these records (the PR was getting big) and maybe some consolidation of records into logical blocks to make it easier for the user (e.g. putting all fission related records in their own class as we did with the metadata).

### CODE REVIEW

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp [hunk 1]
```diff
@@ -0,0 +1,21 @@
...
+
+/**
+ *  @brief Default constructor
+ */
+FissionNeutronSpectrumMatrix( FissionType type ) : RealListRecord( base::Keyword( "chi_mat", type ) ) {}
+
+/**
+ *  @brief Constructor
+ *
+ *  @param[in] values    the fission neutron spectrum matrix values
```
Comment (tgsaller) [inline src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp:14]: Mention type?

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp [hunk 2]
```diff
@@ -0,0 +1,21 @@
+/**
+ *  @brief Default constructor
+ */
+FissionNeutronSpectrumMatrix() : RealListRecord( base::Keyword( "chi_mat" ) ) {}
+
+/**
+ *  @brief Default constructor
```
Comment (tgsaller) [inline src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp:7]: Mention type?

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/verify.hpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
+/**
+ *  @brief Verify the fission neutron spectrum vector values
+ *
+ *  The following verification tests are performed:
+ *    - there is at least one value
+ *
```
Comment (tgsaller) [inline src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/verify.hpp:6]: Mention the "check for matrix" test.

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/test/FissionNeutronSpectrumMatrix.test.cpp [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+        chunk.print( output );
+
+        CHECK( buffer == record );
+      } // THEN
+    } // WHEN
+  } // GIVEN
+
+  GIVEN( "invalid data for a FissionNeutronSpectrumMatrix instance" ) {
+
+    WHEN( "the number of weight values is insufficient" ) {
```
Comment (tgsaller) [inline src/NDItk/multigroup/FissionNeutronSpectrumMatrix/test/FissionNeutronSpectrumMatrix.test.cpp:81]: Not weights.

Comment (tgsaller) [reply src/NDItk/multigroup/FissionNeutronSpectrumMatrix/test/FissionNeutronSpectrumMatrix.test.cpp:81]: Also, we should test the "matrix not square" bit too.

File: src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp [hunk 5]
```diff
@@ -0,0 +1,21 @@
+/**
+ *  @brief Default constructor
+ */
+FissionNeutronSpectrumVector() : RealListRecord( base::Keyword( "chi_vec" ) ) {}
+
+/**
+ *  @brief Default constructor
```
Comment (tgsaller) [inline src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp:7]: Type?

File: src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+ */
+FissionNeutronSpectrumVector( FissionType type ) : RealListRecord( base::Keyword( "chi_vec", type ) ) {}
+
+/**
+ *  @brief Constructor
+ *
+ *  @param[in] values    the fission neutron spectrum vector values
```
Comment (tgsaller) [inline src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp:14]: Type?

File: python/src/multigroup/FissionNeutronProduction.python.cpp [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+  using FissionType = njoy::NDItk::multigroup::FissionType;
+
+  // wrap views created by this record
+
+  // create the record
+  python::class_< Record > record(
+
+    module,
+    "FissionNeutronProduction",
+    "A fission neutron production record for multigroup neutron and photon data"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronProduction.python.cpp:29]: COnsider it done. See fix/review-part6.

File: python/src/multigroup/FissionNeutronProduction.python.cpp [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+  // wrap the record
+  record
+  .def(
+
+    python::init< FissionType, std::vector< double > >(),
+    python::arg( "type" ), python::arg( "values" ),
+    "Initialise the record\n\n"
+    "Arguments:\n"
+    "    self      the record\n"
+    "    values    the fission neutron production values"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronProduction.python.cpp:41]: Changed in fix/review-part6.

File: python/src/multigroup/FissionNeutronMultiplicity.python.cpp [hunk 9]
```diff
@@ -0,0 +1,73 @@
...
+    "from_string",
+    [] ( const std::string& string, unsigned int number ) -> Record
+       { return readWithFissionSubtype< Record >( string, number ); },
+    python::arg( "string" ), python::arg( "number" ),
+    "Read the record from a string\n\n"
+    "An exception is raised if something goes wrong while reading the\n"
+    "record\n\n"
+    "Arguments:\n"
+    "    string    the string representing the record\n"
+    "    number    the fission neutron multiplicity values"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronMultiplicity.python.cpp:66]: Yes. Changed in fix/review-part6.

File: python/src/multigroup/FissionNeutronProduction.python.cpp [hunk 10]
```diff
@@ -0,0 +1,73 @@
...
+    "from_string",
+    [] ( const std::string& string, unsigned int number ) -> Record
+       { return readWithFissionSubtype< Record >( string, number ); },
+    python::arg( "string" ), python::arg( "number" ),
+    "Read the record from a string\n\n"
+    "An exception is raised if something goes wrong while reading the\n"
+    "record\n\n"
+    "Arguments:\n"
+    "    string    the string representing the record\n"
+    "    number    the fission neutron production values"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronProduction.python.cpp:66]: Yes. Changed in fix/review-part6.

File: python/src/multigroup/FissionNeutronSpectrumMatrix.python.cpp [hunk 11]
```diff
@@ -0,0 +1,80 @@
...
+  using FissionType = njoy::NDItk::multigroup::FissionType;
+
+  // wrap views created by this record
+
+  // create the record
+  python::class_< Record > record(
+
+    module,
+    "FissionNeutronSpectrumMatrix",
+    "A fission neutron spectrum matrix record for multigroup neutron and photon data"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronSpectrumMatrix.python.cpp:29]: I changed it everywhere. Changed in fix/review-part6.

File: python/src/multigroup/FissionNeutronSpectrumMatrix.python.cpp [hunk 12]
```diff
@@ -0,0 +1,80 @@
...
+    "from_string",
+    [] ( const std::string& string, unsigned int number ) -> Record
+       { return readWithFissionSubtype< Record >( string, number ); },
+    python::arg( "string" ), python::arg( "number" ),
+    "Read the record from a string\n\n"
+    "An exception is raised if something goes wrong while reading the\n"
+    "record\n\n"
+    "Arguments:\n"
+    "    string    the string representing the record\n"
+    "    number    the number of groups"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronSpectrumMatrix.python.cpp:73]: The matrix is square.

File: python/src/multigroup/FissionNeutronSpectrumVector.python.cpp [hunk 13]
```diff
@@ -0,0 +1,73 @@
...
+  using FissionType = njoy::NDItk::multigroup::FissionType;
+
+  // wrap views created by this record
+
+  // create the record
+  python::class_< Record > record(
+
+    module,
+    "FissionNeutronSpectrumVector",
+    "A fission neutron spectrum vector record for multigroup neutron and photon data"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronSpectrumVector.python.cpp:29]: Changed in fix/review-part6.

File: python/src/multigroup/FissionNeutronSpectrumVector.python.cpp [hunk 14]
```diff
@@ -0,0 +1,73 @@
...
+    "from_string",
+    [] ( const std::string& string, unsigned int number ) -> Record
+       { return readWithFissionSubtype< Record >( string, number ); },
+    python::arg( "string" ), python::arg( "number" ),
+    "Read the record from a string\n\n"
+    "An exception is raised if something goes wrong while reading the\n"
+    "record\n\n"
+    "Arguments:\n"
+    "    string    the string representing the record\n"
+    "    number    the number of groups"
```
[the remark this answers is not in the document]
Comment (whaeck) [reply python/src/multigroup/FissionNeutronSpectrumVector.python.cpp:66]: Yes, leaving this as is.

File: src/NDItk/MultigroupTable.hpp [hunk 15]
```diff
@@ -128,6 +145,78 @@ class MultigroupTable {
...
+  fissionNeutronMultiplicity( const multigroup::FissionType& type ) const {
+
+    switch ( type ) {
+
+      case multigroup::FissionType::Prompt  : return this->nubar_prompt_;
+      case multigroup::FissionType::Delayed : return this->nubar_delayed_;
+      case multigroup::FissionType::Total   : return this->nubar_total_;
+      default : {
+
+        throw std::runtime_error( "This code is unreachable, contact a developer" );
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable.hpp:161]: This should never happen but a compiler can complain about missing a default case.

File: src/NDItk/MultigroupTable/src/verify.hpp [hunk 16]
```diff
@@ -87,32 +87,35 @@ void verify() {
...
-  if ( types > 0 ) {
+  if ( this->metadata().numberOutgoingParticles().has_value() ) {
 
-    if ( ( this->outgoingParticleTypes().numberOutgoingParticles() != types ) ||
-         ( this->outgoingParticleTransportData().numberOutgoingParticles() != types ) ||
-         ( this->outgoing_structure_.size() != types ) ||
-         ( this->outgoing_production_.size() != types ) ||
-         ( this->outgoing_heating_.size() != types ) ||
-         ( this->outgoing_kerma_.size() != types ) ) {
+    const auto types = this->metadata().numberOutgoingParticles().value();
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/src/verify.hpp:92]: changed to number in fix/review-part6.

File: src/NDItk/MultigroupTable/src/verify.hpp [hunk 17]
```diff
@@ -87,32 +87,35 @@ void verify() {
...
-                 this->outgoing_structure_.size() );
-      Log::info( "Number of types in the outgoing production matrices: {}",
-                 this->outgoing_production_.size() );
-      Log::info( "Number of types in the outgoing heating numbers: {}",
-                 this->outgoing_heating_.size() );
-      Log::info( "Number of types in the outgoing kerma values: {}",
-                 this->outgoing_kerma_.size() );
-      throw std::exception();
+      if ( ( this->outgoingParticleTypes().numberOutgoingParticles() != types ) ||
+           ( ! this->outgoingParticleTransportData().empty() && this->outgoingParticleTransportData().numberOutgoingParticles() != types ) ||
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/MultigroupTable/src/verify.hpp:96]: No. the records are independent so someone could have omitted one type on the outgoingParticleTypes() record and/or given more data on the outgoingParticleTransportData() record.

File: src/NDItk/multigroup/FissionNeutronMultiplicity/test/FissionNeutronMultiplicity.test.cpp [hunk 18]
```diff
@@ -0,0 +1,129 @@
...
+        chunk.print( output );
+
+        CHECK( buffer == record );
+      } // THEN
+    } // WHEN
+  } // GIVEN
+
+  GIVEN( "invalid data for a FissionNeutronMultiplicity instance" ) {
+
+    WHEN( "the number of weight values is insufficient" ) {
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronMultiplicity/test/FissionNeutronMultiplicity.test.cpp:75]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronProduction/src/ctor.hpp [hunk 19]
```diff
@@ -0,0 +1,21 @@
+/**
+ *  @brief Default constructor
+ */
+FissionNeutronProduction() : RealListRecord( base::Keyword( "nu_sig_f" ) ) {}
+
+/**
+ *  @brief Default constructor
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronProduction/src/ctor.hpp:7]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronProduction/src/ctor.hpp [hunk 20]
```diff
@@ -0,0 +1,21 @@
...
+ */
+FissionNeutronProduction( FissionType type ) : RealListRecord( base::Keyword( "nu_sig_f", type ) ) {}
+
+/**
+ *  @brief Constructor
+ *
+ *  @param[in] values    the fission neutron production values
```
[the remark this answers is not in the document]
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronProduction/src/ctor.hpp:14]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp (hunk 1 above)
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp:14]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp (hunk 2 above)
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/ctor.hpp:7]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/verify.hpp (hunk 3 above)
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumMatrix/src/verify.hpp:6]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronSpectrumMatrix/test/FissionNeutronSpectrumMatrix.test.cpp (hunk 4 above)
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumMatrix/test/FissionNeutronSpectrumMatrix.test.cpp:81]: Damn you copy/paste!

Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumMatrix/test/FissionNeutronSpectrumMatrix.test.cpp:81]: Changed in fix/review-part6. And added square test.

File: src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp (hunk 5 above)
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp:7]: Changed in fix/review-part6.

File: src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp (hunk 6 above)
Comment (whaeck) [reply src/NDItk/multigroup/FissionNeutronSpectrumVector/src/ctor.hpp:14]: Changed in fix/review-part6.

**Pull request merged**